scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -0,0 +1,344 @@
+      import CoGrouped._
+
+      def pipeToCG[V1](t: TypedPipe[(K, V1)]): CoGroupable[K, V1] =
+        t match {
+          case ReduceStepPipe(cg: CoGroupable[K @unchecked, V1 @unchecked]) =>
+            // we are relying on the fact that we use Ordering[K]
+            // as a contravariant type, despite it not being defined
+            // that way.
+            cg
+          case CoGroupedPipe(cg) =>
+            // we are relying on the fact that we use Ordering[K]
+            // as a contravariant type, despite it not being defined
+            // that way.
+            cg.asInstanceOf[CoGroupable[K, V1]]
+          case kvPipe => IdentityReduce(cg.keyOrdering, kvPipe, None, Nil)
+        }
+
+      cg match {
+        case p@Pair(_, _, _) =>
+          def go[A, B, C](pair: Pair[K, A, B, C]): LiteralPipe[(K, C)] = {
+            val llit = handleCoGrouped(pair.larger, recurse)
+            val rlit = handleCoGrouped(pair.smaller, recurse)
+            val fn = pair.fn
+            Binary(llit, rlit, { (l: TypedPipe[(K, A)], r: TypedPipe[(K, B)]) =>
+              Pair(pipeToCG(l), pipeToCG(r), fn)
+            })
+          }
+          widen(go(p))
+        case wr@WithReducers(_, _) =>
+          def go[V1 <: V](wr: WithReducers[K, V1]): LiteralPipe[(K, V)] = {
+            val reds = wr.reds
+            Unary[TypedPipe, (K, V1), (K, V)](handleCoGrouped(wr.on, recurse), { (tp: TypedPipe[(K, V1)]) =>
+              tp match {
+                case ReduceStepPipe(rs) =>
+                  withReducers(rs, reds)
+                case CoGroupedPipe(cg) =>
+                  // we are relying on the fact that we use Ordering[K]
+                  // as a contravariant type, despite it not being defined
+                  // that way.
+                  CoGroupedPipe(WithReducers(cg, reds))
+                case kvPipe =>
+                  ReduceStepPipe(IdentityReduce(cg.keyOrdering, kvPipe, None, Nil)
+                    .withReducers(reds))
+              }
+            })
+          }
+          widen(go(wr))
+        case wd@WithDescription(_, _) =>
+          def go[V1 <: V](wd: WithDescription[K, V1]): LiteralPipe[(K, V)] = {
+            val desc = wd.description
+            Unary[TypedPipe, (K, V1), (K, V)](handleCoGrouped(wd.on, recurse), { (tp: TypedPipe[(K, V1)]) =>
+              tp match {
+                case ReduceStepPipe(rs) =>
+                  withDescription(rs, desc)
+                case CoGroupedPipe(cg) =>
+                  // we are relying on the fact that we use Ordering[K]
+                  // as a contravariant type, despite it not being defined
+                  // that way.
+                  CoGroupedPipe(WithDescription(cg, desc))
+                case kvPipe =>
+                  kvPipe.withDescription(desc)
+              }
+            })
+          }
+          widen(go(wd))
```

Comment:
Some of these widens might not be needed, since the `go` method is returning `LiteralPipe[(K, V)]` already.